app/Models/BeatmapSet.php

```diff
@@ -169,63 +171,67 @@ public function isQualified()
         return $this->approved == static::QUALIFIED;
     }
 
-    public static function search(array $params = [])
+    private static function sanitizeSearchParams(array &$params = [])
     {
-
-        // default search params
-        $params += [
-            'query' => null,
-            'mode' => 0,
-            'status' => 0,
-            'genre' => null,
-            'language' => null,
-            'extra' => null,
-            'rank' => null,
-            'page' => 1,
-            'sort' => ['ranked', 'desc'],
-        ];
-        extract($params);
-
-        if (count($sort) != 2) {
-            $sort = ['ranked', 'desc'];
+        // sort param
+        if (count($params['sort']) != 2) {
+            $params['sort'] = ['ranked', 'desc'];
         }
 
-        list($sort_field, $sort_order) = $sort;
-
         $valid_sort_fields = ['title', 'artist', 'creator', 'difficulty', 'ranked', 'rating', 'plays'];
-
-        if (!in_array($sort_field, $valid_sort_fields) || !in_array($sort_order, ['asc', 'desc'])) {
-            $sort_field = 'ranked';
-            $sort_order = 'desc';
+        $valid_sort_orders = ['asc', 'desc'];
+        if (!in_array($params['sort'][0], $valid_sort_fields) || !in_array($params['sort'][1], $valid_sort_orders)) {
+            $params['sort'] = ['ranked', 'desc'];
         }
 
-        // remap fields to their db/elastic-search equivalents
-        $sort_field = str_replace(
+        // remap sort field to their db/elastic-search equivalents
+        $params['sort'][0] = str_replace(
             ['difficulty', 'ranked', 'plays'],
             ['difficultyrating', 'approved_date', 'playcount'],
-            $sort_field
+            $params['sort'][0]
         );
 
-        $max = config('osu.beatmaps.max', 50);
+        list($params['sort_field'], $params['sort_order']) = $params['sort'];
+        unset($params['sort']);
+
+        $valid_ranks = ['A', 'B', 'C', 'D', 'S', 'SH', 'X', 'XH'];
+        foreach ($params['rank'] as $rank) {
+            if (!in_array($rank, $valid_ranks)) {
+                unset($params['rank'][$rank]);
+            }
+        }
+
+        if ($params['query'] !== null) {
+            $params['query'] = preg_replace('/\s{2,}/', ' ', $params['query']);
+            $params['query'] = trim($params['query']);
+
+            $query_parts = explode(' ', $params['query']);
```

Comment:
If you're talking about using regex instead of explode, [obligatory joke about using regex](http://www.xkcd.com/1171/). Also [this](http://micro-optimization.com/explode-vs-preg_split.html) and [this](http://stackoverflow.com/questions/27303235/in-php-which-is-faster-preg-split-or-explode) >_>

FWIW I haven't really bothered optimising stuff yet. just getting it all going first~